---
show: step
version: 1.0
enable_checker: true
---

# 主控程序

## 回忆上次内容

- 上次把 apple.py 拆分成了
	- 输入模块
	- 主模块
- 引用模块中变量的时候
	- 要带上模块(module)的名字空间(namespace)
	  - get_fruits.a
	  - get_fruits.b
- 最终 
	- 拆分代码 成功！

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20230529-1685370373197)

- 可以将程序 再拆分 吗？
- 拆成
	- 输入模块
	- 输出模块
- 然后 
	- 再由主控模块
	- 进行调用
	- 可以吗？🤔
- 什么是主控程序呢？


### 主控模块

- 就像一个核心的中控室

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20230618-1687046321895)

- 就像主神经中枢
	- 是一切汇总的地方
	- 管理着具体的步骤

### 建立主控


```
vi -o main.py get_fruits.py output.py
``` 

- 新建一个 main.py 文件
	- 导入 get_fruits、output 两个包
	- 然后运行

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20221030-1667097849653)

- 增大窗口宽度
	- <kbd>ctrl</kbd>+<kbd>w</kbd> , <kbd>+</kbd> 
- 减小窗口宽度
	- <kbd>ctrl</kbd>+<kbd>w</kbd> , <kbd>-</kbd> 

### 运行

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20221030-1667097869196)

- 运行成功

### 程序结构

- 既然我在 main.py 里面
	- 已经导入了 get_fruits 这个包

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220214-1644820630458)

- 我还有必要在 output 中
	- 再次声明导入这个包么？

### 实践

- 实践出真知
	- 试着删除output中的第一句

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220520-1653052173153)

### 尝试

- 失败！

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220214-1644820720723)

- 还是需要在 output 中导入 get_fruits 这个包的
	- 虽然 main 认识了 get_fruits
	- 但是 output 不认识也不行
	- 不认识 就无法得到输入的数值
- 我们跳出具体文件
	- 去观察文件夹结构

### 文件结构

- 可以看到
	- 出现了两个被引用包的 pyc 文件

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220520-1653052296468)

- 位于`__pycache__`文件夹中

### 顺序过程

- 其实被引用的功能代码 
	- 可以分成三块
	  - 输入
	  - 处理
	  - 输出

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20210816-1629054611474)

- 这是一个典型的 `面向过程`的 程序
  - `面向过程`就是 Procedure Oriented
  - 按部就班 
	- 把复杂的过程
		- 分解成 相对简单明确的模块
    - 每个模块 做自己简单明确的事情
    - 专业分工
    - 提高效率
- 这是工业时代
	- 提高效率的思维模式

### 工业化

> 以福特公司为代表的建立在流水线分工基础上的劳动组织方式和大批量生产模式，极大地提高了劳动生产率，也部分地提高了工人的工资水平和福利待遇，从而使整个社会的消费水平有了大幅度提高。

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20230614-1686749636264)
> Workers in a Ford Motor Company factory construct a Model T engine on an assembly line about 1914.

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20230614-1686749648241)

- 分工细化
	- 造就了 
		- 学科体系
		- 摩登时代


### 分工明确

- 铁路警察
	- 各管一段
- 不要想着怎么发明轮子
	- 轮子有现成的
	- 拿来用就完事了

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20211020-1634736291247)

- 接下来
	- 我们要把模块分成三块

| 功能 | 作用 | 模块名 |
| --- | --- | --- |
| 输入 | 接收到什么信号 |  get_fruits |
| 处理 | 怎么处理信号 | process |
| 输出 | 最终输出什么信号 | output |


### 具体步骤

- 三步走

| get_fruits | process | output|
| --- | --- | --- |
| 输入水果数量| 对水果数量求和 | 输出最终结果 |

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220214-1644822881677)

- main 
	- 主控程序
	- 需要导入 3 个 包
	- 然后依次执行
- get_fruits 
	- 输入模块
	- 虽然被main和process导入了两次
	- 但是 输入数值的部分
		- 只被执行一次
- process 
	- 导入 get_fruits 模块后
	- 进行对 模块中的a和b 
		- 进行求和操作
- output 
	- 不需要导入 get_fruits 模块
	- 导入 process 模块 
		- 得到 求和的结果 process.total 

### 测试结果

- 数字(123)当做测试用例的话
	- 能成功

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20230530-1685457620607)

- 但是用字母(abc)当测试用例的话
	- 会如何呢？

### 不成功 🙅🏻‍♀️

- 问题出在process.py中
	- int(a) + int(b)那一步

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20221030-1667098894283)

- 所以需要在那里
	- try 
	- except
- 试着将a转化为整数
	- 如果转化失败 
	- 则进行报错

### 处理异常

- 修改代码

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220520-1653052973851)

### 运行结果

- 字母的测试用例
	- 可以顺利地报错

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20220520-1653053001159)

- 但是现在还面临一个问题
	- 就是不明确
	- 不知道
		- 是 apple数量 出了问题
		- 还是 banana数量 出了问题
	- 报错细节不明确

### 修改代码

- 在process模块中
	- 对于a和b `分别`进行转整数操作	
	- `分别`try...except...
	- 如果a 转化中出了错误
		- 则报错a 不能转化为整数
	- 如果b 转化中出了错误
		- 则报错b 不能转化为整数

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20221030-1667099566219)

- 这样运行可以吗？

### 优化代码

- 这样运行是没有问题的
- 不过应该还有更好的方式

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20221030-1667099616015)

- 数据类型转化工作
	- 应该 可以在输入部分 完成
- 输入变量的 容错处理
	- 应该在输入的 时候 
		- 就第一时间做
	- 不要留到 数据处理时 才做

## 总结

- 我们细化了主控程序(main.py)
	- 将程序分成三个具体步骤

| get_fruits | process | output|
| --- | --- | --- |
| 输入水果数量| 对水果数量求和 | 输出最终结果 |

![图片描述](https://doc.shiyanlou.com/courses/uid1190679-20230710-1688949895944)

- 使用了 try 结构
  - try
  - except
  - 发现异常就报错
- 输入部分 的 数据类型错误
  - 应该在 输入部分 来报错
- 具体怎么报呢？🤔
- 下次再说！👋
